### PR TITLE
[13.0][FIX] vault: Prevent 200 records limit in search panels.

### PR DESCRIPTION
**Pull request by victoralmau**
Changes done:
- [x] Prevent 200 records limit in search panels.
- [x] Show in the left sidebar only the records that have children.

This is quite similar to https://github.com/OCA/dms/pull/167

Please @pedrobaeza and @CarlosRoca13 can you review it?

@Tecnativa TT38731

### CODE REVIEW

File: vault/models/vault_entry.py [hunk 1]
```diff
@@ -1,17 +1,21 @@
...
+from odoo import _, api, fields
 from odoo.exceptions import ValidationError
 
+from odoo.addons.web.models import models
+
 _logger = logging.getLogger(__name__)
 
 
-class VaultEntry(models.Model):
+class VaultEntry(odoo.models.Model):
```
Comment (pedrobaeza) [inline vault/models/vault_entry.py]: I insist: why this change?

**Comment by pedrobaeza** [comment]
Not sure we should do this feature change. @fkantelberg what do yo think?

File: vault/models/vault_entry.py (hunk 1 above)
Comment (CarlosRoca13) [reply vault/models/vault_entry.py]: It's because he is adding this models: https://github.com/OCA/server-auth/pull/411/files#diff-40d91c6c9961012399719d24970edab4a33704ffb9d72e74ba0528c9b966f378R13

Comment (pedrobaeza) [reply vault/models/vault_entry.py]: Then better to do `from odoo.addons.web.models import models as web_models`

File: vault/tests/test_vault.py [hunk 2]
```diff
@@ -155,3 +159,13 @@ def test_search_expired(self):
         self.assertTrue(domain[0] == "|")
         self.assertIn(("expire_date", "=", False), domain)
         self.assertTrue(any(("expire_date", ">=") == d[:2] for d in domain))
+
+    def test_vaul_entry_search_panel_limit(self):
```
Comment (fkantelberg) [inline vault/tests/test_vault.py]: Typo in `vault`

File: vault/models/vault_entry.py [hunk 3]
```diff
@@ -71,10 +75,25 @@ def _compute_complete_name(self):
...
+        """Add context to show just the name in searchpanel instead of full path.
+        We need to override function to prevent 200 records limit in search panel."""
+        old_value = models.SEARCH_PANEL_LIMIT
+        models.SEARCH_PANEL_LIMIT = False
+        _self = self.with_context(
+            search_panel_domain_override=True, entry_short_name=True
+        )
+        res = super(VaultEntry, _self).search_panel_select_range(field_name)
+        models.SEARCH_PANEL_LIMIT = old_value
+        return res
```
Comment (fkantelberg) [inline vault/models/vault_entry.py]: I get the meaning behind it but the fact that you already did that here in multiple locations sounds like a required core patch instead of a hack like this in every location.

I would rather suggest to patch odoo.addons.web.models.models like this:
```
search_panel_limit = self.env.context.get("search_panel_limit", SEARCH_PANEL_LIMIT)
return {
    'parent_field': parent_name,
    'values': Comodel.with_context(hierarchical_naming=False).search_read([], fields, limit=search_panel_limit),
}
```
and set the context here.

File: vault/models/vault_entry.py [hunk 4]
```diff
@@ -71,10 +75,25 @@ def _compute_complete_name(self):
...
+
+    def search_read(self, domain=None, fields=None, offset=0, limit=None, order=None):
+        """We apply a domain to not list records that do not have children."""
+        if self.env.context.get("search_panel_domain_override"):
+            domain = [("child_ids", "!=", False)]
```
Comment (fkantelberg) [inline vault/models/vault_entry.py]: From a navigational perspective this makes sense to not show the leaves of the tree. I'm not quite sure if you want to fully overwrite it or extend the domain here. It could make a difference if your view/filter is set.

**Review by fkantelberg — CHANGES_REQUESTED** [review]
Thank you for your suggestions. See my comments. For large databases a change like this makes absolutely sense.

File: vault/models/vault_entry.py (hunk 1 above)
Comment (victoralmau) [reply vault/models/vault_entry.py]: Finally, this is not necessary.

File: vault/models/vault_entry.py (hunk 3 above)
Comment (victoralmau) [reply vault/models/vault_entry.py]: Finally, this is not necessary.

I have changed the approach in another way to simplify code.

**Comment by victoralmau** [comment]
Changes done.

**Review by fkantelberg — APPROVED** [review]
Looks less hacky now :+1:

**Comment by pedrobaeza** [comment]
Please explain a bit in the commit message and code comment what leads to that implementation.

**Comment by victoralmau** [comment]
Changes done.

**Review by pedrobaeza — APPROVED** [review]
